File: c++/src/liquidplanner_demo.cpp
#include <iostream>
#include <termios.h>

#include "liquidplanner.h"

/*
 * LiquidPlanner C++ Api-Example
 *
 * See liquidplanner.cpp & h for examples of adding new requests.
 * See models/model.h for for the base Model.
 * Use models/person.cpp & h as a pattern for adding additional models.
 */

using namespace std;

template <typename Type> void print(Type ts) {
  for (typename Type::iterator i = ts.begin(); i < ts.end(); i++) {
    cout << "\t" << i->name() << endl;
  }
}

void echoOff();
void echoOn();

int main ()
{
  string username;
  string password;

  cout << "Enter username: ";
  cin >> username;

  echoOff();

  cout << "Enter password: ";
  cin >> password;

  echoOn();

  cout << endl;

//------------------------------------------------------------------------------
//- connect to LiquidPlanner
//------------------------------------------------------------------------------

  LiquidPlanner lp(username, password);

//------------------------------------------------------------------------------
//- get current user
//------------------------------------------------------------------------------

  Person account = lp.getAccount();
  if (account.hasError()) { 
    cout << "there was an error " << account.error() 
         << "check your username and password" << endl;
    return 1;
  } else {
    cout << account.fullName() << endl;
  }

//------------------------------------------------------------------------------
//- list spaces and select the first space
//------------------------------------------------------------------------------

  Workspaces spaces = account.workspaces();
  cout << endl << "Workspaces:" << endl;
  print<Workspaces>(spaces);
   
  Workspace firstSpace = spaces[spaces.size()-1];
  lp.setSpace(firstSpace.id());
  cout << endl << "Using: " << firstSpace.name() << endl;
  
//------------------------------------------------------------------------------
//- list projects
//------------------------------------------------------------------------------

  Projects firstSpaceProjects = lp.getProjects();
  cout << endl << firstSpace.name() << " tasks:" << endl;
  print<Projects>(firstSpaceProjects);

//------------------------------------------------------------------------------
//- create a new task
//------------------------------------------------------------------------------

  Json::Value taskParams;
  taskParams["name"] = "Learn the API";

  Task task = lp.createTask(taskParams);

//------------------------------------------------------------------------------
//- list tasks
//------------------------------------------------------------------------------

  Tasks firstSpaceTasks = lp.getTasks();
  cout << endl << "After create: ";
  cout << firstSpace.name() << " tasks:" << endl;
  print<Tasks>(firstSpaceTasks);
  
//------------------------------------------------------------------------------
//- update the task by renaming it
//- and setting its owner_id to the id of the account fetched above
//------------------------------------------------------------------------------

  taskParams["name"] = "update the api example";
  taskParams["owner_id"] = account.id();

  lp.updateTask(task, taskParams);

//------------------------------------------------------------------------------
//- and list tasks
//------------------------------------------------------------------------------

  firstSpaceTasks = lp.getTasks();
  cout << endl << "After update: ";
  cout << firstSpace.name() << " tasks:" << endl;
  print<Tasks>(firstSpaceTasks);

//------------------------------------------------------------------------------
//- update the task to clean up 
//------------------------------------------------------------------------------

  lp.deleteTask(task);

//------------------------------------------------------------------------------
//- and list tasks
//------------------------------------------------------------------------------

  firstSpaceTasks = lp.getTasks();
  cout << endl << "After delete: ";
  cout << firstSpace.name() << " tasks:" << endl;
  print<Tasks>(firstSpaceTasks);

//------------------------------------------------------------------------------
//- that's all folks! 
//------------------------------------------------------------------------------

  return 0;
}


//------------------------------------------------------------------------------
//- helpers to mute keystrokes when entering the password
//------------------------------------------------------------------------------

void echoOff() {
  termios tty;

  tcgetattr(STDIN_FILENO, &tty);

  tty.c_lflag &= ~ECHO;

  tcsetattr(STDIN_FILENO, TCSANOW, &tty);
}

void echoOn() {
  termios tty;

  tcgetattr(STDIN_FILENO, &tty);

  tty.c_lflag |= ECHO;

  tcsetattr(STDIN_FILENO, TCSANOW, &tty);
}

File: go/liquidplanner_demo.go
package main

import (
  "bufio"
  "fmt"
  "os"
  "strings"

  "code.google.com/p/gopass"

  lp "github.com/liquidplanner/liquidplanner_go"
)

func main() {
  promptForEmailAndPassword();

//------------------------------------------------------------------------------
//-- Fetch your account
//------------------------------------------------------------------------------

  person, response := lp.GetAccount()

  if response.Error != nil {
    fmt.Println( "There was an error. Check your email and password?" );
    os.Exit(1);
  }

  fmt.Println( "Account:" )
  fmt.Println( "\t" + person.FullName() )

//------------------------------------------------------------------------------
//-- List your workspaces
//------------------------------------------------------------------------------

  workspaces := person.Workspaces

  fmt.Println( "\nWorkspaces:" )

  for _,space := range workspaces {
    fmt.Println( "\t" + space.Name )
  }

//------------------------------------------------------------------------------
//-- Select the first space
//------------------------------------------------------------------------------

  workspace := workspaces[len(workspaces)-1]

  lp.SetSpace( workspace )

//------------------------------------------------------------------------------
//-- List Projects 
//------------------------------------------------------------------------------

  projects := lp.GetProjects()

  fmt.Println( "\n" + workspace.Name + " projects:" )

  for _,project := range projects {
    fmt.Println( "\t" + project.Name )
  }

//------------------------------------------------------------------------------
//-- Create a task
//------------------------------------------------------------------------------

  fmt.Println( "\n-- Create a Task")

  task := lp.CreateTask(lp.Task {
    Name: "Learn the API", 
  })

  fmt.Println( "\n" + workspace.Name + " tasks:" )

  for _,task := range lp.GetTasks() {
    fmt.Println( "\t" + task.Name )
  }

//------------------------------------------------------------------------------
//-- Update the task
//------------------------------------------------------------------------------

  fmt.Println( "\n-- Update the Task")

  task.Name = "update with the API"

  lp.UpdateTask(&task)

  fmt.Println( "\n" + workspace.Name + " tasks:" )

  for _,task := range lp.GetTasks() {
    fmt.Println( "\t" + task.Name )
  }

//------------------------------------------------------------------------------
//-- Delete task
//------------------------------------------------------------------------------

  fmt.Println( "\n-- Delete the Task")

  lp.DeleteTask( &task )

  fmt.Println( "\n" + workspace.Name + " tasks:" )

  for _,task := range lp.GetTasks() {
    fmt.Println( "\t" + task.Name )
  }

//------------------------------------------------------------------------------
//-- That's all folks!
//------------------------------------------------------------------------------

}





func promptForEmailAndPassword() {
  reader := bufio.NewReader(os.Stdin)

  // get username and password
  fmt.Print( "Enter email: " )
  user,_ := reader.ReadString('\n')

  pass,_ := gopass.GetPass("Enter Password: ")

  user = strings.TrimSuffix(user, "\n")
  pass = strings.TrimSuffix(pass, "\n")

  lp.Login(user,pass)
}
